src/gt4py/next/ffront/signature.py

```diff
@@ -0,0 +1,138 @@
+# Free Software Foundation, either version 3 of the License, or any later
+# version. See the LICENSE.txt file at the top-level directory of this
+# distribution for a copy of the license or check <https://www.gnu.org/licenses/>.
+#
+# SPDX-License-Identifier: GPL-3.0-or-later
+
+from __future__ import annotations
+
+import functools
+import inspect
+import types
+from typing import Any, Callable
+
+from gt4py.next.ffront import (
+    field_operator_ast as foast,
+    program_ast as past,
+    stages as ffront_stages,
+)
+from gt4py.next.type_system import type_specifications as ts
+
+
+def should_be_positional(param: inspect.Parameter) -> bool:
+    return (param.kind is inspect.Parameter.POSITIONAL_ONLY) or (
+        param.kind is inspect.Parameter.POSITIONAL_OR_KEYWORD
+    )
+
+
+@functools.singledispatch
+def make_signature(func: Any) -> inspect.Signature:
+    """Make a signature for a Python or DSL callable, which suffices for use in 'convert_to_positional'."""
+    if isinstance(func, types.FunctionType):
+        return inspect.signature(func)
+    raise NotImplementedError(f"'make_signature' not implemented for {type(func)}.")
+
+
+@make_signature.register(foast.ScanOperator)
+@make_signature.register(past.Program)
+@make_signature.register(foast.FieldOperator)
+def signature_from_fieldop(func: foast.FieldOperator) -> inspect.Signature:
+    if isinstance(func.type, ts.DeferredType):
+        raise NotImplementedError(
+            f"'make_signature' not implemented for pre type deduction {type(func)}."
+        )
+    fieldview_signature = func.type.definition
+    return inspect.Signature(
+        parameters=[
+            inspect.Parameter(name=str(i), kind=inspect.Parameter.POSITIONAL_ONLY)
+            for i, param in enumerate(fieldview_signature.pos_only_args)
+        ]
+        + [
+            inspect.Parameter(name=k, kind=inspect.Parameter.POSITIONAL_OR_KEYWORD)
+            for k in fieldview_signature.pos_or_kw_args
+        ],
+    )
+
+
+@make_signature.register(ffront_stages.FieldOperatorDefinition)
+def signature_from_fieldop_def(func: ffront_stages.FieldOperatorDefinition) -> inspect.Signature:
+    signature = make_signature(func.definition)
+    if func.node_class == foast.ScanOperator:
+        return inspect.Signature(list(signature.parameters.values())[1:])
+    return signature
+
+
+@make_signature.register(ffront_stages.ProgramDefinition)
+def signature_from_program_def(func: ffront_stages.ProgramDefinition) -> inspect.Signature:
+    return make_signature(func.definition)
+
+
+@make_signature.register(ffront_stages.FoastOperatorDefinition)
+def signature_from_foast_stage(func: ffront_stages.FoastOperatorDefinition) -> inspect.Signature:
+    return make_signature(func.foast_node)
+
+
+@make_signature.register
+def signature_from_past_stage(func: ffront_stages.PastProgramDefinition) -> inspect.Signature:
+    return make_signature(func.past_node)
+
+
+def convert_to_positional(
```

Comment:
@tehrengruber I would also particularly like your opinion on this. Do you think it could / should be merged with something else? Do you see a corner case where this might fail (which needs to be covered with more tests)?